zebra-network/src/peer_set/candidate_set.rs

```diff
@@ -3,67 +3,80 @@ use std::sync::{Arc, Mutex};
-/// │  └─────────────┘              ╲ ╱                     │
-/// │         │                      V                      │
-/// │         │disconnected_peers    │                      │
-/// │         ▼                      │                      │
-/// │         Λ     filter by        │                      │
-/// │        ╱ ╲ !contains_addr      │                      │
-/// │       ▕   ▏◀───────────────────┼──────────────────────┤
-/// │        ╲ ╱                     │                      │
-/// │         V                      │                      │
-/// │         │                      │                      │
-/// │┌────────┼──────────────────────┼──────────────────────┼────────┐
-/// ││        ▼                      ▼                      │        │
-/// ││ ┌─────────────┐        ┌─────────────┐        ┌─────────────┐ │
-/// ││ │Disconnected │        │  Gossiped   │        │Failed Peers │ │
-/// ││ │    Peers    │        │    Peers    │        │ AddressBook │◀┼┐
-/// ││ │ AddressBook │        │ AddressBook │        │             │ ││
-/// ││ └─────────────┘        └─────────────┘        └─────────────┘ ││
-/// ││        │                      │                      │        ││
-/// ││ #1 drain_oldest        #2 drain_newest        #3 drain_oldest ││
-/// ││        │                      │                      │        ││
-/// ││        ├──────────────────────┴──────────────────────┘        ││
-/// ││        │           disjoint candidate sets                    ││
-/// │└────────┼──────────────────────────────────────────────────────┘│
-/// │         ▼                                                       │
-/// │         Λ                                                       │
-/// │        ╱ ╲         filter by                                    │
-/// └──────▶▕   ▏!is_potentially_connected                            │
-///          ╲ ╱                                                      │
-///           V                                                       │
+///             filter by            Λ
+///          !contains_addr         ╱ ╲
+///  ┌────────────────────────────▶▕   ▏
+///  │                              ╲ ╱
+///  │                               V
+///  │                               │
+///  │                               │
+///  │                               │
+///  │                               │
+///  │                               │
+///  │                               │
+///  │                               │
+///  │                               │
+///  ├───────────────────────────────┼───────────────────────────────┐
+///  │ PeerSet AddressBook           ▼                               │
+///  │ ┌─────────────┐       ┌────────────────┐      ┌─────────────┐ │
+///  │ │  Possibly   │       │`NeverAttempted`│      │  `Failed`   │ │
+///  │ │Disconnected │       │     Peers      │      │   Peers     │◀┼┐
+///  │ │  `Recent`   │       │                │      │             │ ││
```

Comment:
Split `Recent` into `Live` and `Dead`, transition peers to `Dead` on every query.
Send all queries through a single function that does the `Dead` updates.